samples/cpp/CMakeLists.txt

```diff
@@ -1,117 +1,65 @@
-# ----------------------------------------------------------------------------
-#  CMake file for C samples. See root CMakeLists.txt
-#
-# ----------------------------------------------------------------------------
-
-SET(OPENCV_CPP_SAMPLES_REQUIRED_DEPS opencv_core opencv_imgproc opencv_flann
-    opencv_imgcodecs opencv_videoio opencv_highgui opencv_ml opencv_video
-    opencv_objdetect opencv_photo opencv_features2d opencv_calib3d
-    opencv_stitching opencv_videostab opencv_shape ${OPENCV_MODULES_PUBLIC} ${OpenCV_LIB_COMPONENTS})
-
+ocv_install_example_src(cpp *.cpp *.hpp CMakeLists.txt)
+
+set(OPENCV_CPP_SAMPLES_REQUIRED_DEPS
+  opencv_core
+  opencv_imgproc
+  opencv_flann
+  opencv_imgcodecs
+  opencv_videoio
+  opencv_highgui
+  opencv_ml
+  opencv_video
+  opencv_objdetect
+  opencv_photo
+  opencv_features2d
+  opencv_calib3d
+  opencv_stitching
+  opencv_videostab
+  opencv_shape
+  ${OPENCV_MODULES_PUBLIC}
+  ${OpenCV_LIB_COMPONENTS})
 ocv_check_dependencies(${OPENCV_CPP_SAMPLES_REQUIRED_DEPS})
 
+if(NOT BUILD_EXAMPLES OR NOT OCV_DEPENDENCIES_FOUND)
+  return()
+endif()
 
-if(BUILD_EXAMPLES AND OCV_DEPENDENCIES_FOUND)
-  project(cpp_samples)
-
-  ocv_include_directories("${OpenCV_SOURCE_DIR}/include")#for opencv.hpp
-  ocv_include_modules_recurse(${OPENCV_CPP_SAMPLES_REQUIRED_DEPS})
-
-  if(HAVE_opencv_cudaoptflow)
-    ocv_include_directories("${OpenCV_SOURCE_DIR}/modules/cudaoptflow/include")
+project(cpp_samples)
+ocv_include_modules_recurse(${OPENCV_CPP_SAMPLES_REQUIRED_DEPS})
+file(GLOB_RECURSE cpp_samples RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.cpp)
+if(NOT HAVE_OPENGL)
+  ocv_list_filterout(cpp_samples Qt_sample)
+endif()
+if(NOT HAVE_opencv_cudaarithm OR NOT HAVE_opencv_cudafilters)
+  ocv_list_filterout(cpp_samples "/gpu/")
+endif()
+if(NOT VTK_USE_FILE)
```

Comment:
This condition should check "opencv_viz" module build status too

See this [build failure](http://pullrequest.opencv.org/buildbot/builders/master-mac-static/builds/10368)